src/compose/app.ts

```diff
@@ -513,8 +594,15 @@ export class App {
 			networkPairs: Array<ChangingPair<Network>>;
 			volumePairs: Array<ChangingPair<Volume>>;
 			servicePairs: Array<ChangingPair<Service>>;
+			appsToLock: AppsToLockMap;
 		} & UpdateState,
 	): Nullable<CompositionStep> {
+		const currentServicesLocked = this.services.every((svc) =>
+			context.locksTaken.isLocked(svc.appId, svc.serviceName),
+		);
+		const targetServicesLocked = context.targetApp.services.every((svc) =>
+			context.locksTaken.isLocked(svc.appId, svc.serviceName),
+		);
```

Comment:
Again, I'm not sure we need to distinguish between current and target services locked. The app either has its services locked or not, if there are new services, `takeLock` should preemptively take those locks

Reply:
I combined `currentServicesLocked` and `targetServicesLocked` in commit f199319d0a2ecf3dcbfc59c82f022b662d784a71. Any lock-taking calculations will also generate steps for both current & target services now, in that same commit